alpheios_nemo_ui/__init__.py

```diff
@@ -534,6 +535,147 @@ def new_level(self,subreference,prev):
+                subend = substart
+
+            # now loop through the precaculated references and try to find one that
+            # the custom reference is part of
+            for index,reff in enumerate(reffs):
+                reff = Reference(reff)
+                refstart = reff.start
+                refend = reff.end
+                # these should all be ranges but if not just make the end the start
+                if refend is None:
+                    refend = reff.start
+                if refstart == substart:
+                    # custom reference starts the exact beginning of
+                    # a precalculated reference
+                    # so the previous precalculated reference is retained whole
+                    # and the next reference is the immediate next sibling to the
+                    #  end of the precalculated range
+                    nextRange = str(nextPassage) + '-' + str(refend)
+                    if index > 0:
+                        prevRange = reffs[index-1]
+                    else:
+                        prevRange = None
+                    return prevRange, nextRange
+                elif refend == subend:
+                    # custom reference ends at the exact end of the precalculated
+                    # reference so the previous reference is the beginning
+                    # of the precalculated reference to the previous sibling and
+                    # the next reference is retained whole
+                    prevRange = str(refstart) + "-" + str(prevPassage)
+                    if index < len(reffs) - 1:
+                        nextRange = reffs[index+1]
+                    else:
+                        nextRange = None
+                    return prevRange, nextRange
+                else:
+                    # check to see if the custom reference is somewhere in the middle
+                    latest_start = self.latest_reff(refstart, substart)
+                    latest_end = self.latest_reff(refend,subend)
+                    if latest_start == substart and latest_end == refend:
+                        # custom reference is fully enclosed in the precalculated reference
+                        # previous reference is beginning of the precalculated reference
+                        # to the previous sibling and next reference is the next sibling to the
+                        # end of the precaculated reference
+                        return str(refstart) + "-" + str(prevPassage), str(nextPassage) + '-' + str(refend)
+        # if no matching range could be found, fall back to just the immeidate siblings as next and previous
+        return passage.siblingsId
+
+    def latest_reff(self, refA, refB):
+        """ Calculate which reff is the later one in the text
+            2.1.4 is later than 2.1
+            2.2 is later than 2.1.4
+            2.1.610 is later than 2.1.61
+        """
+        try:
+            lengthA = len(refA)
```

Comment:
I think that is essentially what I am doing. The function Reference.start.list takes Reference and breaks it into its dot-separated parts. (Although all of our texts right now do use a dot, in fact, it does not have to be a dot that separates the parts, which is why I am trying to reuse the library code here to create the list rather than parsing it myself)

But you might be right, the three different tests I have here (different length citations, equal length and equal parent and different parent) might be able to combined into one set of tests. Will look at this again. Thank you!

Reply:
Well, I tried streamlining this code further, and it broke one of the scenarios, so I think I'm going to leave it as is. I am sure it could be made more concise, but I am out of time to do that right now :-(